### ISSUE TITLE
jump on uninitialised value introduced by commit b25a103638bc5111

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12806]
Migrated from [rt.perl.org#116899](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=116899) (status was 'resolved')

Searchable as RT116899$

**Comment by p5pRT** [comment]
### From @nwc10

#### Created by @nwc10



This commit&#8203;:

commit b25a103638bc5111446c1e65a21db00a6e13acc1
Author&#8203;: Karl Williamson \<public@&#8203;khwilliamson\.com>
Date&#8203;:   Thu Dec 27 12&#8203;:34&#8203;:41 2012 \-0700

&nbsp;    regcomp\.c&#8203;: Clean up ANYOF\_CLASS handling\.

&nbsp;    The ANYOF\_CLASS flag is used in ANYOF nodes \(for \[bracketed\] and the
&nbsp;    synthetic start class\) only when matching something like \\w\, \[&#8203;:punct&#8203;:\]
&nbsp;    etc\.\, under /l \(locale\)\.  It should not be set unless /l is specified\.
&nbsp;    However\, it was always getting set for the synthetic start class\.  This
&nbsp;    commit fixes that\.  The previous code was masking errors in which it was
&nbsp;    being tested for unnecessarily\, and for much of the 5\.17 series\, the
&nbsp;    synthetic start class was always set to test for locale\, which was a
&nbsp;    waste of cpu when no locale was specified\.


unfortunately introduces this error&#8203;:

$ valgrind \./miniperl \-Ilib \-e '"" =~ /\\x\{100\}/i;'
==22821== Memcheck\, a memory error detector
==22821== Copyright \(C\) 2002\-2011\, and GNU GPL'd\, by Julian Seward et al\.
==22821== Using Valgrind\-3\.7\.0 and LibVEX; rerun with \-h for copyright info
==22821== Command&#8203;: \./miniperl \-Ilib \-e ""\\ =~\\ /\\\\x\{100\}/i;
==22821== 
==22821== Conditional jump or move depends on uninitialised value\(s\)
==22821==    at 0x469E58&#8203;: S\_cl\_is\_anything \(regcomp\.c&#8203;:824\)
==22821==    by 0x479862&#8203;: Perl\_re\_op\_compile \(regcomp\.c&#8203;:6199\)
==22821==    by 0x4171A2&#8203;: Perl\_pmruntime \(op\.c&#8203;:4611\)
==22821==    by 0x463A1E&#8203;: Perl\_yyparse \(perly\.y&#8203;:1330\)
==22821==    by 0x513E18&#8203;: S\_doeval \(pp\_ctl\.c&#8203;:3445\)
==22821==    by 0x5162B8&#8203;: Perl\_pp\_require \(pp\_ctl\.c&#8203;:4026\)
==22821==    by 0x4C234C&#8203;: Perl\_runops\_standard \(run\.c&#8203;:42\)
==22821==    by 0x4088FB&#8203;: Perl\_call\_sv \(perl\.c&#8203;:2707\)
==22821==    by 0x557B58&#8203;: Perl\_\_core\_swash\_init \(utf8\.c&#8203;:3015\)
==22821==    by 0x55674E&#8203;: Perl\_to\_utf8\_case \(utf8\.c&#8203;:2406\)
==22821==    by 0x555250&#8203;: Perl\_\_to\_uni\_fold\_flags \(utf8\.c&#8203;:1844\)
==22821==    by 0x485FB2&#8203;: S\_regatom \(regcomp\.c&#8203;:10876\)
==22821== 
==22821== 
==22821== HEAP SUMMARY&#8203;:
==22821==     in use at exit&#8203;: 1\,545\,277 bytes in 9\,964 blocks
==22821==   total heap usage&#8203;: 40\,466 allocs\, 30\,502 frees\, 3\,940\,476 bytes allocated
==22821== 
==22821== LEAK SUMMARY&#8203;:
==22821==    definitely lost&#8203;: 0 bytes in 0 blocks
==22821==    indirectly lost&#8203;: 0 bytes in 0 blocks
==22821==      possibly lost&#8203;: 324\,728 bytes in 104 blocks
==22821==    still reachable&#8203;: 1\,220\,549 bytes in 9\,860 blocks
==22821==         suppressed&#8203;: 0 bytes in 0 blocks
==22821== Rerun with \-\-leak\-check=full to see details of leaked memory
==22821== 
==22821== For counts of detected and suppressed errors\, rerun with&#8203;: \-v
==22821== Use \-\-track\-origins=yes to see where uninitialised values come from
==22821== ERROR SUMMARY&#8203;: 16 errors from 1 contexts \(suppressed&#8203;: 4 from 4\)


Line 824 is the if\(\) here&#8203;:

/\* Can match anything \(initialization\) \*/
STATIC int
S\_cl\_is\_anything\(const struct regnode\_charclass\_class \*cl\)
\{
&nbsp;    int value;

&nbsp;    PERL\_ARGS\_ASSERT\_CL\_IS\_ANYTHING;

&nbsp;    for \(value = 0; value \< ANYOF\_MAX; value \+= 2\)
&nbsp;    if \(ANYOF\_CLASS\_TEST\(cl\, value\) && ANYOF\_CLASS\_TEST\(cl\, value \+ 1\)\)
&nbsp;        return 1;
&nbsp;    if \(\!\(cl\->flags & ANYOF\_UNICODE\_ALL\)\)
&nbsp;    return 0;
&nbsp;    if \(\!ANYOF\_BITMAP\_TESTALLSET\(\(const void\*\)cl\)\)
&nbsp;    return 0;
&nbsp;    return 1;
\}


This is pretty much all I know\.

Nicholas Clark

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl 5.17.10:

Configured by nick at Thu Feb 21 20:49:16 CET 2013.

Summary of my perl5 (revision 5 version 17 subversion 10) configuration:
  Derived from: 34d6d453862a4b2db2b03735dbe1083b9b659d24
  Platform:
    osname=linux, osvers=2.6.32-5-amd64, archname=x86_64-linux
    uname='linux gcc20 2.6.32-5-amd64 #1 smp mon jan 16 16:22:28 utc 2012 x86_64 gnulinux '
    config_args='-Dusedevel=y -Dcc=ccache gcc -Dld=gcc -Dcf_email=nick@ccl4.org -Dperladmin=nick@ccl4.org -Dinc_version_list=  -Dinc_version_list_init=0 -Doptimize=-g -Uusethreads -Uuselongdouble -Uuse64bitall -Uusemymalloc -Duseperlio -Dprefix=~/Sandpit/snap-v5.17.9-17-g34d6d45 -Uusevendorprefix -Uvendorprefix=~/Sandpit/snap-v5.17.9-17-g34d6d45 -Uuserelocatableinc -Umad -Uuseshrplib -de'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='ccache gcc', ccflags ='-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-g',
    cppflags='-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.4.5', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib/../lib /usr/lib/../lib /lib /usr/lib /lib64 /usr/lib64
    libs=-lnsl -ldl -lm -lcrypt -lutil -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.11.3.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.11.3'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -g -L/usr/local/lib -fstack-protector'

Locally applied patches:
    


@INC for perl 5.17.10:
    lib
    /home/nick/Sandpit/snap-v5.17.9-17-g34d6d45/lib/perl5/site_perl/5.17.10/x86_64-linux
    /home/nick/Sandpit/snap-v5.17.9-17-g34d6d45/lib/perl5/site_perl/5.17.10
    /home/nick/Sandpit/snap-v5.17.9-17-g34d6d45/lib/perl5/5.17.10/x86_64-linux
    /home/nick/Sandpit/snap-v5.17.9-17-g34d6d45/lib/perl5/5.17.10
    .


Environment for perl 5.17.10:
    HOME=/home/nick
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/nick/Perl/rakudo/install/bin:/home/nick/bin:/opt/local/bin:/opt/local/sbin:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games:/usr/local/sbin:/sbin:/usr/sbin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @khwilliamson

Note that this problem doesn't manifest if \-DDEBUGGING is specified\,
which is unfortunate
\-\- 
Karl Williamson

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @khwilliamson

Fixed by commit c96939e471cb3942b61ec7b103b0449a9fde922d
\-\- 
Karl Williamson

**Comment by p5pRT** [comment]
@khwilliamson - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @nwc10

On Fri\, Feb 22\, 2013 at 12&#8203;:31&#8203;:21PM \-0800\, Karl Williamson via RT wrote&#8203;:
> Fixed by commit c96939e471cb3942b61ec7b103b0449a9fde922d

Cool\. Thanks\.

Nicholas Clark

**Comment by p5pRT** [comment]
### From @khwilliamson

Actually\, the commit that introduced the bug was mine from 3 weeks
earlier\, 9cf78fa04cfee93425003b87b547dfa158d8fe9a


\-\- 
Karl Williamson
